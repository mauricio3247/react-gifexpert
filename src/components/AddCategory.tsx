import React, { useState } from 'react'
import PropTypes from 'prop-types'
type CategoryProperties = {
    setCategories:  React.Dispatch<React.SetStateAction<string[]>>
}

export const AddCategory = ({setCategories}: CategoryProperties) => {

    const [inputValue, setInputValue] = useState('')
    const handleInput = (e:any) => {
        setInputValue(e.target.value)
    }

    const handleSubmit = (e: any) => {
        e.preventDefault();
        if (inputValue.trim().length > 2) {
            setCategories((categories)=> [inputValue, ...categories])
            setInputValue('')
        }
        
    }
    return (
        <form onSubmit= {handleSubmit}>
            <input 
            onChange={handleInput} 
            value={inputValue} />
        </form>
    )
}


AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}