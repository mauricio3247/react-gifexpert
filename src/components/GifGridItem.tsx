import React from 'react'
import PropTypes from 'prop-types'
export const GifGridItemComponent = ({title, url, id}:any) => {
    return (
        <div className="card animate__animated animate__zoomIn">
            <img src={url} alt={title} ></img>
            <p>{title}</p>
        </div>
    )
}

GifGridItemComponent.propTypes = {
    title: PropTypes.string.isRequired, 
    url: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired
}