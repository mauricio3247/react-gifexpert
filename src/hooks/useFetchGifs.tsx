import { useEffect, useState } from "react"
import { getGifs } from "../helpers/getGifs"

type stateFetchGift = {
    data: Array<any>,
    loading: boolean
}

export const useFetchGifs = (category:string) => {
    const [state, setstate] = useState<stateFetchGift>({
        data: [],
        loading: true
    })

    useEffect(()=> {
        getGifs(category)
        .then(images => setstate ({
            data: images,
            loading: false
        }))
    }, [category])


    return state;
}