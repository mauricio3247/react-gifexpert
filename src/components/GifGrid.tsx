import React from 'react'
import PropTypes from 'prop-types'
import { GifGridItemComponent } from './GifGridItem'

import { useFetchGifs } from '../hooks/useFetchGifs'

type GifGridProperties = {
    category: string
}

export const GifGridComponent = ({category}: GifGridProperties)=> {
    
    const {data:images, loading} = useFetchGifs(category);

    return (
        <>
        <h3 className="animate__animated animate__fadeIn">{category}</h3>
        <div className='card-grid'>
            
            { loading && <p className="animate__animated animate__flash">'Loading...'</p> }

            {
                images.map ( img => (
                    <GifGridItemComponent 
                    {...img}
                    key={img.id} />
                ))
            }
                

        </div>
        </>
    )
}

GifGridComponent.propTypes = {
    category: PropTypes.string.isRequired
}