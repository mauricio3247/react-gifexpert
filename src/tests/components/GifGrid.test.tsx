import React from 'react'
import { shallow, ShallowWrapper } from 'enzyme'
import { GifGridComponent } from '../../components/GifGrid'


/**
 * sustituimos la dependencia por una funcion mocked que automaticamente genere los valores esperados
 */
import { useFetchGifs } from '../../hooks/useFetchGifs'
jest.mock('../../hooks/useFetchGifs')
const useFetchGifsMocked = useFetchGifs as jest.MockedFunction<typeof useFetchGifs>;


describe('test componente GifGridComponent', () => {

    const category= 'One Punch'


    test('Debe crearse el componente', ()=> {
        useFetchGifsMocked.mockReturnValue({
            data: [],
            loading: true
        })
        const wrapper = shallow (<GifGridComponent category={category} />)
        expect (wrapper).toMatchSnapshot()
    })


    test('Debe mostrar items al cargar imagenes', () => {
        const gifs = [
            {
                id: '123',
                title: 'Un titulo',
                url: 'https://localhost/algo.jpg'
            }
        ]

        useFetchGifsMocked.mockReturnValue({
            data: gifs,
            loading: false
        })

        const wrapper = shallow (<GifGridComponent category={category} />)
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('p').exists()).toBe(false)
        expect(wrapper.find('GifGridItemComponent').length).toBe(gifs.length)

    })

})
