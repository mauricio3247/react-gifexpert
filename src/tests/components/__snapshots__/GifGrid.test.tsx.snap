// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test componente GifGridComponent Debe crearse el componente 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    One Punch
  </h3>
  <div
    className="card-grid"
  >
    <p
      className="animate__animated animate__flash"
    >
      'Loading...'
    </p>
  </div>
</Fragment>
`;

exports[`test componente GifGridComponent Debe mostrar items al cargar imagenes 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    One Punch
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItemComponent
      id="123"
      key="123"
      title="Un titulo"
      url="https://localhost/algo.jpg"
    />
  </div>
</Fragment>
`;
